### PR TITLE
Feature/yoast payload optimize

### LINKED ISSUE #563: Reduce Yoast overhead by removing and preventing additional queries for Yoast metadata
### Is your enhancement related to a problem? Please describe.

We recently did some work to reduce the Yoast payload and optimize hydration but there's still room for improving performance by preventing Yoast from doing a lot of additional queries on WP by disabling these additional queries in the Plugin.

We don't need every post to have an associated `yoast_head` and `yoast_head_json`. We only need the first post and the associated queried object to have Yoast metadata.

Bellow are examples of different scenarios and which entities would need the yoast metadata.

# Scenarios

## Scenario 1

If we're querying for `{ postType: 'post', slug: 'hello-world' }`. Only the first post needs Yoast metadata, There's no queriedObjected on this request so metadata on associated terms can be removed as well.

## Scenario 2

If we're querying for `{ postType: 'post', per_page: 10, category: 'news' }`, only the first post and the first term of the `news` category needs the yoast metadata.

## Scenario 3

If we're querying for `{ postType: 'post', per_page: 10, author: 'jane' }`, only the first post and the jane author needs the yoast metadata.

## Scenario 4

If we're querying for `{ postType: 'post', per_page: 10, category: 'news', author: 'jane' }`, only the first post. the jane author needs the first term of the `news` category needs the yoast metadata.

# Implementation

We can activate this optimization via `headless.config.js`.

```js
module.exports = {
	integrations: {
		yoastSEO: {
			enable: true,
                        optimizeYoastPayload: true,
		},
	},
}
```

Then if `optimizeYoastPayload` is true our fetch strategies would send a `optimizeYoastPayload` query param to all requests which would enable the plugin code to optimize yoast metadata. This allows the optimization to be controlled by the front-end versus having to activate via a filter/setting in WP.

### Designs

_No response_

### Describe alternatives you've considered

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by lucymtc**
### Description of the Change

Optimize Yoast payload response ensuring only necessary Yoast data is added to the elements in _embedded that are needed.

Closes https://github.com/10up/headstartwp/issues/563

### Changelog Entry

> Added - New feature
> Changed - Existing functionality
> Deprecated - Soon-to-be removed feature
> Removed - Feature
> Fixed - Bug fix
> Security - Vulnerability

### Credits

Props @username, @username2, ...

### Checklist:

- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: wp/headless-wp/tests/php/tests/TestYoastIntegration.php [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	 * @return void
+	 */
+	public function test_optimise_category_yoast_payload() {
+
+		// Perform a REST API request for the posts by category.
+		$result = $this->get_posts_by_with_optimised_response( 'categories', $this->category_id );
+		$this->assert_yoast_head_in_response( $result );
+
+		// Perform a REST API request for the posts by author.
+		// $result = $this->get_posts_by_with_optimised_response( 'author', $this->author_id );
```
Comment (lucymtc) [inline wp/headless-wp/tests/php/tests/TestYoastIntegration.php]: @nicholasio Not sure why this request returns empty data although posts are created with the correct author id

File: wp/headless-wp/includes/classes/Integrations/YoastSEO.php [hunk 2]
```diff
@@ -322,4 +326,126 @@ function ( $presenters ) {
...
+	 *
+	 * @param array            $result The response data to be served, typically an array.
+	 * @param \WP_REST_Server  $server Server instance.
+	 * @param \WP_REST_Request $request Request used to generate the response.
+	 *
+	 * @return array Modified response data.
+	 */
+	public function optimise_yoast_payload( $result, $server, $request, $embed = false ) {
+
+		$embed = $embed ?: rest_parse_embed_param( $_GET['_embed'] ?? false );
```
Comment (lucymtc) [inline wp/headless-wp/includes/classes/Integrations/YoastSEO.php]: $embed added to be able to force it for unit tests.

I still need to add the check for the param sent from the app when adding optimizeYoastPayload and integrate it in the nextjs

File: wp/headless-wp/includes/classes/Integrations/YoastSEO.php [hunk 3]
```diff
@@ -322,4 +326,127 @@ function ( $presenters ) {
...
+
+			if ( ! empty( $post_obj['_embedded']['wp:term'] ) ) {
+				$this->optimise_yoast_payload_for_taxonomy( $post_obj['_embedded']['wp:term'], $request, $first_post );
+			}
+
+			if ( ! empty( $post_obj['_embedded']['author'] ) ) {
+				$this->optimise_yoast_payload_for_author( $post_obj['_embedded']['author'], $request, $first_post );
+			}
+
+			if ( ! $first_post ) {
```
Comment (nicholasio) [inline wp/headless-wp/includes/classes/Integrations/YoastSEO.php]: There's one edge case: if the request contains `slug` it might return multiple results withmultiple posts sharing the same slug but be under different parents, and that is resolved on the front-end via post path matching. 

I think if slug is set we should not unset yoast_head and yoast_head_json for any of the results, so I think you can simply check if slug it not set here.

**Review by nicholasio — COMMENTED** [review]
This is looking good! Just left one comment about a potential edge case, can you add a test for when `slug` is passed and it might return multiple posts with different parents.

Also please add a changeset for `@headstartwp/core` and `@headstartwp/headstartwp`

File: wp/headless-wp/includes/classes/Integrations/YoastSEO.php (hunk 3 above)
Comment (lucymtc) [reply wp/headless-wp/includes/classes/Integrations/YoastSEO.php]: I have pushed the change @nicholasio also pushed a change to set 'up' embeddable as false so it doesn't attempt to get the data for  parent posts https://github.com/10up/headstartwp/pull/727/commits/246fed907cbae71a37765f8c3b5388661a39ffe0

**Comment by nicholasio** [comment]
@lucymtc can you also handle `wp:featuredmedia`

**Review by nicholasio — APPROVED** [review]
